/* This patch file is from Qualcomm Technologies, Inc. and is provided under the following license:
 * Copyright (c) Qualcomm Technologies, Inc. and/or its subsidiaries.
 * SPDX-License-Identifier: BSD-3-Clause-Clear
 */
From 0f157d0cafc996931a4e5b2cd1610fe6f311a07c Mon Sep 17 00:00:00 2001
From: Valerie Cagle <dcagle@qti.qualcomm.com>
Date: Fri, 7 Nov 2025 14:24:59 -0800
Subject: [PATCH] Revert "Teach switch_root to move mountpoints into the new
 root."

This reverts commit c04b565204eb6b7e3508ac8dd42539ab97752635.

Signed-off-by: Valerie Cagle <dcagle@qti.qualcomm.com>
---
 toys/other/switch_root.c | 54 +++++++++++++++++++---------------------
 1 file changed, 25 insertions(+), 29 deletions(-)

diff --git a/toys/other/switch_root.c b/toys/other/switch_root.c
index 1c75060..bd778fc 100644
--- a/toys/other/switch_root.c
+++ b/toys/other/switch_root.c
@@ -19,36 +19,24 @@ config SWITCH_ROOT
 
 #define FOR_switch_root
 #include "toys.h"
+#include <sys/vfs.h>
 
 GLOBALS(
   char *c;
 
-  struct stat new;
   dev_t rootdev;
 )
 
 static int del_node(struct dirtree *node)
 {
-  int flag = 0;
-
-  if (same_file(&TT.new, &node->st) || !dirtree_notdotdot(node)) return 0;
-
-  if (node->st.st_dev != TT.rootdev) {
-    char *s = dirtree_path(node, 0);
-
-    if (mount(s, s+1, "", MS_MOVE, "")) perror_msg("Failed to move %s", s);
-    // TODO: handle undermounts
-    rmdir(s);
-    free(s);
-
-    return 0;
-  }
-
-  if (S_ISDIR(node->st.st_mode)) {
-    if (!node->again) return DIRTREE_COMEAGAIN;
-    flag = AT_REMOVEDIR;
+  if (node->st.st_dev == TT.rootdev && dirtree_notdotdot(node)) {
+    int flag = 0;
+    if (S_ISDIR(node->st.st_mode)) {
+      if (!node->again) return DIRTREE_COMEAGAIN;
+      flag = AT_REMOVEDIR;
+    }
+    unlinkat(dirtree_parentfd(node), node->name, flag);
   }
-  unlinkat(dirtree_parentfd(node), node->name, flag);
 
   return 0;
 }
@@ -56,12 +44,13 @@ static int del_node(struct dirtree *node)
 void switch_root_main(void)
 {
   char *newroot = *toys.optargs, **cmdline = toys.optargs+1;
-  struct stat st;
+  struct stat st1, st2;
   struct statfs stfs;
-  int ii, console QUIET;
+  int console QUIET;
 
-  // Must be root on a ramfs or tmpfs instance
   if (getpid() != 1) error_exit("not pid 1");
+
+  // Root filesystem we're leaving must be ramfs or tmpfs
   if (statfs("/", &stfs) ||
     (stfs.f_type != 0x858458f6 && stfs.f_type != 0x01021994))
   {
@@ -70,13 +59,13 @@ void switch_root_main(void)
   }
 
   // New directory must be different filesystem instance
-  if (chdir(newroot) || stat(".", &TT.new) || stat("/", &st) ||
-    same_file(&TT.new, &st))
+  if (chdir(newroot) || stat(".", &st1) || stat("/", &st2) ||
+    st1.st_dev == st2.st_dev)
   {
     error_msg("bad newroot '%s'", newroot);
     goto panic;
   }
-  TT.rootdev = st.st_dev;
+  TT.rootdev=st2.st_dev;
 
   // trim any / characters from the init cmdline, as we want to test it with
   // stat(), relative to newroot. *cmdline is also used below, but by that
@@ -84,7 +73,7 @@ void switch_root_main(void)
   while (**cmdline == '/') (*cmdline)++;
 
   // init program must exist and be an executable file
-  if (stat(*cmdline, &st) || !S_ISREG(st.st_mode) || !(st.st_mode&0100)) {
+  if (stat(*cmdline, &st1) || !S_ISREG(st1.st_mode) || !(st1.st_mode&0100)) {
     error_msg("bad init");
     goto panic;
   }
@@ -97,6 +86,12 @@ void switch_root_main(void)
   // Ok, enough safety checks: wipe root partition.
   dirtree_read("/", del_node);
 
+  // Fix the appearance of the mount table in the newroot chroot
+  if (mount(".", "/", NULL, MS_MOVE, NULL)) {
+    perror_msg("mount");
+    goto panic;
+  }
+
   // Enter the new root before starting init
   if (chroot(".")) {
     perror_msg("chroot");
@@ -110,11 +105,12 @@ void switch_root_main(void)
   }
 
   if (TT.c) {
-    for (ii = 0; ii<3; ii++) dup2(console, ii);
+    int i;
+    for (i=0; i<3; i++) if (console != i) dup2(console, i);
     if (console>2) close(console);
   }
   execv(*cmdline, cmdline);
   perror_msg("Failed to exec '%s'", *cmdline);
 panic:
-  if (FLAG(h)) for (;;) wait(NULL);
+  if (toys.optflags & FLAG_h) for (;;) wait(NULL);
 }
